### DISCUSSION TITLE
Supporting multiple ECR authentication methods.

### DISCUSSION THREAD

**Discussion by jheddings** [renovatebot/renovate#20326] [Suggest an Idea]
### Type of discussion.

I'm proposing an idea

### Tell us more.

AWS provides two authentication methods for accessing ECR.  Currently, renovate supports using AWS credentials to generate an authentication token.  I'm proposing a change to the config that would allow additional ways for configuring ECR access.  Each of these are applied to a `hostRules` entry for hosts matching ECR registries.

Reference for authentication methods: https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html

## AWS Credentials

This is the currently supported method.  Providing `username` [access key ID], `password` [secret key] and (optionally) `token` [session token] uses the AWS API to generate an authorization token.  This is passed as a `Basic` auth header.

## Login Password (basic auth)

In this case, the user has already obtained a login password for the registry (using `aws ecr get-login-password` or similar). 
 By specifying only `password` on the ECR host, it will be used to create a `Basic` auth header with the `AWS` user.

## Authorization Token

For this case, a user has obtained an authorization token for the registry (using `aws ecr get-authorization-token` or similar).  By setting only `token` on the ECR host, it will be passed directly using the `Basic` auth header.

I've started a PR to show these methods...  It's not complete, but I wanted to have it available as part of this discussion - #20320

**Comment by rarkins** [comment]
What's the typical expiry of the last two auth types? e.g. how long until the password or token expires

**Reply by jheddings** [reply]
They follow the same rules as the session token in the AWS credentials...  For our CI, we typically set these to 12 hours, which is the maximum time allowed.

**Reply by rarkins** [reply]
OK, so the use case here is for self-hosted where the token can be generated and inserted into config for each run?

**Reply by jheddings** [reply]
Yes, precisely.  We reference these values using environment variables that are imported from secrets using `existingSecret` in the config.

**Comment by jmuddle-via** [comment]
@jheddings, thank you for this discussion and PR.

I think your proposal would address my discussion #19239 and issue #19241.

**Reply by jheddings** [reply]
Thanks for the follow up...  I still need some time to work out unit tests for the PR, which will hopefully be enough to get it over the line.

**Comment by gabfelp** [comment]
@jheddings thanks for the idea! I think this feature will help me (as many others I saw in discussions)

As context, I'm trying to deploy renovate on my Github Workflow. Before doing so, I assume a role with this AWS action:
```
- name: Authenticate in AWS
  uses: aws-actions/configure-aws-credentials@v3
  with:
    role-to-assume: ${{ env.role_to_assume }}
    aws-region: ${{ env.AWS_REGION }}
```

After that, I change my `$HOME/.docker/config.json` to include

```
	"credHelpers": {
		"<ACCOUNTID>.dkr.ecr.us-east-1.amazonaws.com": "ecr-login",
		"public.ecr.aws": "ecr-login",
	}
```

And this works fine for pulling images from the action/workflow. I'm able to do `docker pull ACCOUNTID.dkr.ecr.us-east-1.amazonaws.com/ANY:ANY` in the following steps.


However, as I don't have an access key ID nor a secret key, the best I can do for "authenticating again" in renovate is:
- use 'AWS' as username with a password (using `aws ecr get-login-password`) [not supported as you pointed out ❌ ]
or 
- Authenticate with a token (`aws ecr get-authorization-token`) [not supported as you pointed out ❌ ]

I also thought something like this would work:
```
        {
            matchHost: "676443572060.dkr.ecr.us-east-1.amazonaws.com",
            token: process.env.ECR_TOKEN,
            authType: "Basic"
        }
```

But the ecrRegex matches this and I'm back to:
```
Using ecr auth for Docker registry
Could not load credentials from any providers (in trace log, ecs.ts line 38)
ECR getAuthorizationToken error (line 39)
Failed to get authHeaders for getTags lookup
```


Anyway, just commenting here to see if you have any workaround for this before this feature comes out

**Reply by gabfelp** [reply]
(created a separate discussion from this comment -- https://github.com/renovatebot/renovate/discussions/25040)
